File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"

#include <math.h>
#include <stdlib.h>

int ability(void) {
    int min = DIE_FACES + 1;
    int sum = 0;

    for (int i = 0; i < ABILITY_ROLLS; i++) {
        int roll = rand() % DIE_FACES + 1;
        int min = roll < min ? roll : min;
        sum += roll;
    }

    return sum - min;
}

int modifier(int score) { return floor((score - 10) / 2.0); }

dnd_character_t make_dnd_character(void) {
    int constitution = ability();

    return (dnd_character_t){.charisma = ability(),
                             .dexterity = ability(),
                             .intelligence = ability(),
                             .strength = ability(),
                             .wisdom = ability(),
                             .constitution = constitution,
                             .hitpoints = modifier(constitution) + 10};
}

File: solutions/python/dnd-character/2/dnd_character.py
from random import randint

ABILITIES = [
    "strength",
    "dexterity",
    "constitution",
    "intelligence",
    "wisdom",
    "charisma",
]


def modifier(constitution):
    return (constitution - 10) // 2


class Character:
    def __init__(self):
        for ability in ABILITIES:
            setattr(self, ability, self.ability())
        self.hitpoints = modifier(self.constitution) + 10

    def ability(self):
        return sum(sorted((randint(1, 6) for _ in range(4)), reverse=True)[:3])
